rng/philox/philox.hpp

```diff
@@ -0,0 +1,202 @@
+    using odd_indices_sequence = typename transform_index_sequence<half_index_sequence, odd_transform>::type;
+
+    static constexpr auto extract_elements = []<std::size_t... Is>(std::index_sequence<Is...>)
+    {
+        constexpr std::array<UIntType, n> temp{consts...};
+        return std::array<UIntType, array_size>{{temp[Is]...}};
+    };
+
+
+public:
+    // types
+    using result_type = UIntType;
+
+    // engine characteristics
+    static constexpr size_t word_size = w;
+    static constexpr size_t word_count = n;
+    static constexpr size_t round_count = r;
+    static constexpr array<result_type, array_size> multipliers = extract_elements(even_indices_sequence{});
+    static constexpr array<result_type, array_size> round_consts = extract_elements(odd_indices_sequence{});
+    static constexpr result_type min() { return 0; }
+    // static constexpr result_type max() { return m - 1; }
+    static constexpr result_type default_seed = 20111115u;
+    // constructors and seeding functions
+    philox_engine() : philox_engine(default_seed) {}
+
+    explicit philox_engine(result_type value) : x{}, k{}, i{n - 1} {
+        k[0] = value & result_mask;
+    }
+
+    template<class Sseq>
+    explicit philox_engine(Sseq& q);
+
+    void seed(result_type value = default_seed);
+    
+    template<class Sseq>
+    void seed(Sseq& q);
+    
+    void set_counter(const array<result_type, n>& counter);
+
+    // equality operators
+    friend bool operator==(const philox_engine& x, const philox_engine& y) {
+        return false;
+    }
+
+    // generating functions
+    result_type operator()()
+    {
+        ++i;
+        if (i == n) {
+            y = philox_generate(k, x); // see below
+            increment_counter();
+            i = 0;
+        }
+        return y[i];
+    }
+
+    void discard(unsigned long long z);
+    // inserters and extractors
+    template<class charT, class traits>
+    friend basic_ostream<charT, traits>&
+    operator<<(basic_ostream<charT, traits>& os, const philox_engine& x);
+
+    template<class charT, class traits>
+    friend basic_istream<charT, traits>&
+    operator>>(basic_istream<charT, traits>& is, philox_engine& x);
+
+
+private: // utilities
+    using uint_types = std::tuple<std::uint8_t, std::uint16_t, std::uint32_t, std::uint64_t>;
+    using promotion_types = std::tuple<std::uint16_t, std::uint32_t, std::uint64_t, __uint128_t>;
+
+    static consteval std::size_t log2(std::size_t val)
+    {
+        return ((val <= 2) ? 1 : 1 + log2(val / 2));
+    }
+
+    static consteval std::size_t ceil_log2(std::size_t val)
+    {
+        std::size_t additive = static_cast<std::size_t>(!std::has_single_bit(val));
+
+        return log2(val) + additive;
+    }
+
+    using counter_type = std::tuple_element_t<ceil_log2(w / CHAR_BIT), uint_types>;
+    using promotion_type = std::tuple_element_t<ceil_log2(w / CHAR_BIT), promotion_types>;
```

Comment:
log2 is frightening...
<bit> header will help you. bit_width() is what you are looking for...

Reply:
we can probably use just a bitset<w> as a counter_type